### PR TITLE
Report progress for persistent data init

### PR DESCRIPTION
**Pull request by CybervisionVolodymyrPogorelov**
This is initial PR to implement the feature. Testing and polishing the code is in progress

### CODE REVIEW

**Comment by CybervisionVolodymyrPogorelov** [comment]
> So, the next step after this change (generating the log file) would be a change to then actually advertise the progress percentage in the kdcluster per-member status. Have you had any experiments/thoughts about that? Intending to do it in this PR or in a subsequent PR?

We will do this as a separate PR. It is better to deliver the feature by small pieces, step by step.

**Comment by joel-bluedata** [comment]
Automated checks are failing on a formatting issue BTW. Try "make format" to see if that will make the necessary change.

What's been the testing on this so far? I'm particularly curious about how it behaves on an app with lots of stuff to copy, e.g. the latest jupyter-notebook in Ezmeral (which does have rsync available). I can try some things on my end but I'm curious what you've already tried.

File: pkg/executor/statefulset.go [hunk 1]
```diff
@@ -679,19 +680,60 @@ func getStartupScript(
...
+        createRsyncLogFileBaseDir,
+        kubedirectorInitProgressLogs, 
+        strings.Join(persistDirs, " "))
+
+    return rsyncCmd
+}
+
+// Generates command that will do copying with cp
+// No way to display progress
+func generateCpCmd(persistDirs []string) string {
```
Comment (joel-bluedata) [inline pkg/executor/statefulset.go:709]: Same things here about function arguments and also the initial blank line.

**Comment by CybervisionVolodymyrPogorelov** [comment]
> Automated checks are failing on a formatting issue BTW. Try "make format" to see if that will make the necessary change.
> 
> What's been the testing on this so far? I'm particularly curious about how it behaves on an app with lots of stuff to copy, e.g. the latest jupyter-notebook in Ezmeral (which does have rsync available). I can try some things on my end but I'm curious what you've already tried.

Thanks for detailed review, I will implement the needed changes. For now, I tested only on simple case: deploy/example_clusters/cr-cluster-centos7-stor.yaml. I will try with more complex input, like latest jupyter-notebook

File: pkg/executor/statefulset.go [hunk 2]
```diff
@@ -679,19 +680,60 @@ func getStartupScript(
...
+// Checks if the rsync command is available. 
+// If rsync is installed the RSYNC_INSTALLED variable will be 0.
+func genrateRsyncInstalledCmd() string {
+    cmd := "rsync -h &>/dev/null; RSYNC_INSTALLED=$?; "
+    return cmd;
+}
+
+// Generates command that will do copying with rsync
+// The progress will be stored in a file.
+func generateRsyncCmd(persistDirs []string) string {
```
[the remark this answers is not in the document]
Comment (CybervisionVolodymyrPogorelov) [reply pkg/executor/statefulset.go]: Done!

File: pkg/executor/statefulset.go (hunk 1 above)
Comment (CybervisionVolodymyrPogorelov) [reply pkg/executor/statefulset.go]: Done!

File: pkg/executor/statefulset.go [hunk 3]
```diff
@@ -679,19 +680,60 @@ func getStartupScript(
 	}
 }
 
+
+// Checks if the rsync command is available. 
+// If rsync is installed the RSYNC_INSTALLED variable will be 0.
```
[the remark this answers is not in the document]
Comment (CybervisionVolodymyrPogorelov) [reply pkg/executor/statefulset.go]: Done!

File: pkg/executor/statefulset.go [hunk 4]
```diff
@@ -679,19 +680,60 @@ func getStartupScript(
...
+    // don't do this copy if the kubedirector.init file already exists in /etc.
+    copyCondition := fmt.Sprintf("! [ -f /mnt%s ]", kubedirectorInit)
+
+    // In order to perform copying rsync will be used. 
+    // It allows to report the progress that will be saved in a file. 
+    // Here we check if the rsync command is installed. 
+    rsyncInstalled := genrateRsyncInstalledCmd()
+    
+    // If the rsync command is not available the cp command will be used.
+    fullCmd := fmt.Sprintf("%s %s && ( [ $RSYNC_INSTALLED != 0 ] && (%s) || (%s)); touch /mnt%s;",
```
[the remark this answers is not in the document]
Comment (CybervisionVolodymyrPogorelov) [reply pkg/executor/statefulset.go]: Done!

File: pkg/executor/statefulset.go [hunk 5]
```diff
@@ -679,19 +680,60 @@ func getStartupScript(
...
+
+// Generates command that will do copying with rsync
+// The progress will be stored in a file.
+func generateRsyncCmd(persistDirs []string) string {
+    
+    // The directory should be created in /mnt in advance,
+    // otherwise the rsync log file will not be created
+    createRsyncLogFileBaseDir := fmt.Sprintf("mkdir -p /mnt%s", filepath.Dir(kubedirectorInitProgressLogs))
+
+    rsyncCmd := fmt.Sprintf("%s; rsync --log-file=/mnt%s --relative -ax %s /mnt;",
```
[the remark this answers is not in the document]
Comment (CybervisionVolodymyrPogorelov) [reply pkg/executor/statefulset.go]: I check the very first version of rsync. The options --relative -a -x are present in all the versions of rsync: https://github.com/WayneD/rsync/blob/v1.6.4/README. In order to be sure that option --log-file is available I added it the rsync status check. The check will fail if the option is not present and cp will be used.

**Review by joel-bluedata — APPROVED** [review]

**Pull request merged**